src/components/userActivity.ts

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2021 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import * as logging from "./logging";
+const log = logging.get("UserActTracker");
+
+interface UserActivityMetadata {
+    /**
+     * The user is active in "private" rooms. Undefined if not.
+     */
+    private?: true;
+    /**
+     * The user was previously active, so we don't have a grace period.
+     */
+    active?: true;
+}
+
+export interface UserActivitySet {
+    users: {[userId: string]: UserActivity};
+}
+
+// eslint-disable-next-line @typescript-eslint/no-namespace,no-redeclare
+export namespace UserActivitySet {
+    export const DEFAULT: UserActivitySet = {
+        users: {}
+    };
+}
+
+export interface UserActivity {
+    ts: number[];
+    metadata: UserActivityMetadata;
+}
+
+export interface UserActivityTrackerConfig {
+    inactiveAfterDays: number;
+    minUserActiveDays: number;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-namespace,no-redeclare
+export namespace UserActivityTrackerConfig {
+    export const DEFAULT: UserActivityTrackerConfig = {
+        inactiveAfterDays: 31,
+        minUserActiveDays: 3,
+    };
+}
+
+export interface UserActivityState {
+    dataSet: UserActivitySet;
+    changed: string[];
+    activeUsers: number;
+}
+type ChangesCallback = (state: UserActivityState) => void;
+
+const ONE_DAY = 24 * 60 * 60 * 1000;
+
+export class UserActivityTracker {
```

Comment:
Bit more doc-string needed here too

Reply:
Added in 6be73b58b0c0c6196c3d098bf8e2d4b651e9ae30